lnd.go

```diff
@@ -630,105 +633,203 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+func NewTLSManager(cfg *Config) *TLSManager {
+	return &TLSManager{
+		cfg: cfg,
+	}
+}
+
+// getConfig returns a TLS configuration for the gRPC server and credentials
 // and a proxy destination for the REST reverse proxy.
-func getTLSConfig(cfg *Config) ([]grpc.ServerOption, []grpc.DialOption,
+func (t *TLSManager) getConfig() ([]grpc.ServerOption, []grpc.DialOption,
 	func(net.Addr) (net.Listener, error), func(), error) {
 
+	tlsCfg, cleanUp, err := t.generateOrRenewCert()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	// Now that we know that we have a ceritificate, let's generate the
+	// required config options.
+	restCreds, err := credentials.NewClientTLSFromFile(
+		t.cfg.TLSCertPath, "",
+	)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	serverCreds := credentials.NewTLS(tlsCfg)
+	serverOpts := []grpc.ServerOption{grpc.Creds(serverCreds)}
+
+	// For our REST dial options, we'll still use TLS, but also increase
+	// the max message size that we'll decode to allow clients to hit
+	// endpoints which return more data such as the DescribeGraph call.
+	// We set this to 200MiB atm. Should be the same value as maxMsgRecvSize
+	// in cmd/lncli/main.go.
+	restDialOpts := []grpc.DialOption{
+		grpc.WithTransportCredentials(restCreds),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(lnrpc.MaxGrpcMsgSize),
+		),
+	}
+
+	// Return a function closure that can be used to listen on a given
+	// address with the current TLS config.
+	restListen := func(addr net.Addr) (net.Listener, error) {
+		// For restListen we will call ListenOnAddress if TLS is
+		// disabled.
+		if t.cfg.DisableRestTLS {
+			return lncfg.ListenOnAddress(addr)
+		}
+
+		return lncfg.TLSListenOnAddress(addr, tlsCfg)
+	}
+
+	return serverOpts, restDialOpts, restListen, cleanUp, nil
+}
+
+// generateOrRenewCert generates a new TLS certificate if we're not using one
+// yet or renews it if it's outdated.
+func (t *TLSManager) generateOrRenewCert() (*tls.Config, func(), error) {
+	// Genereate a TLS pair if we don't have one yet.
+	err := t.createTLSPair()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certData, parsedCert, err := cert.LoadCert(
+		t.cfg.TLSCertPath, t.cfg.TLSKeyPath,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check to see if the certificate needs to be renewed. If it does, we
+	// return the newly generated certificate data instead.
+	certData, err = t.certMaintenance(parsedCert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tlsCfg := cert.TLSConfFromCert(certData)
+
+	cleanUp := t.setUpLetsEncrypt(&certData, tlsCfg)
+
+	return tlsCfg, cleanUp, nil
+}
+
+// If a TLS pair doesn't exist yet, create them and write them to disk.
+func (t *TLSManager) createTLSPair() error {
 	// Ensure we create TLS key and certificate if they don't exist.
-	if !fileExists(cfg.TLSCertPath) && !fileExists(cfg.TLSKeyPath) {
+	if !fileExists(t.cfg.TLSCertPath) && !fileExists(t.cfg.TLSKeyPath) {
```

Comment:
Invert this if so we can just return `nil` if either file exists and have the generation logic non-indented?

Reply:
After playing with this, I think we should keep this part the same, because a couple commits later we add more logic to this function, which seems to make this change not make as much sense imo

Reply:
Oh wait. Nevermind, factoring in another comment you made below, you're approach looks like the right move :+1: :zap: